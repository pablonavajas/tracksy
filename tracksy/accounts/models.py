from django.db import models
from django.contrib.auth.models import User


class Info(models.Model):
    user = models.OneToOneField(User, related_name='info', on_delete=models.CASCADE)
    isStartup = models.BooleanField()


class Connection(models.Model):
    user = models.ForeignKey(User, related_name="connections", on_delete=models.CASCADE)
    owner = models.CharField(max_length=100)
    name = models.CharField(max_length=100, blank=True, null=True)
    url = models.URLField(blank=True, null=True)
    description = models.CharField(max_length=200, blank=True, null=True)

    class Meta:
        unique_together = ['user', 'url']