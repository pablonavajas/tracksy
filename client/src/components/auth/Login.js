import React, { useState, useContext, useEffect } from 'react';
import Footer from '../layout/Footer';

const Login = props => {
  const onSubmit = e => {};
  const onChange = e => {};

  const email = '';
  const password = '';

  return (
    <div>
      <section className="section section-login">
        <div className="form-container">
          <div className="row">
            <div className="col s12 m8 offset-m2 l6 offset-l3">
              <div className="card-panel login blue white-text center">
                <h2>Tracksy Login</h2>
                <form action="index.html" onSubmit={onSubmit}>
                  <div className="input-field">
                    <i className="material-icons prefix">email</i>
                    <input
                      type="email"
                      id="email"
                      onChange={onChange}
                      required
                    />
                    <label className="white-text" for="email">
                      Email
                    </label>
                  </div>
                  <div className="input-field">
                    <i className="material-icons prefix">lock</i>
                    <input
                      type="password"
                      id="password"
                      onChange={onChange}
                      required
                      minLength="6"
                    />
                    <label className="white-text" for="password">
                      Password
                    </label>
                  </div>
                  <input
                    type="submit"
                    value="Login"
                    className="btn btn-large btn-extended grey lighten-4 black-text"
                  />
                </form>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>

    // <div className="form-container">
    //   <h1>
    //     Account <span className="text-primary">Login</span>
    //   </h1>

    //   <form onSubmit={onSubmit}>
    //     <div>
    //       <label htmlFor="email">Email Address</label>
    //       <input
    //         type="email"
    //         name="email"
    //         //value={email}
    //         onChange={onChange}
    //         required
    //       />
    //     </div>
    //     <div>
    //       <label htmlFor="password">Password</label>
    //       <input
    //         type="password"
    //         name="password"
    //         //value={password}
    //         onChange={onChange}
    //         required
    //         minLength="6"
    //       />
    //     </div>
    //     <input
    //       type="submit"
    //       value="Login"
    //       className="btn btn-primary btn-block"
    //     />
    //   </form>
    // </div>
  );
};

export default Login;
