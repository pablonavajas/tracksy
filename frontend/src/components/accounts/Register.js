import React, { useState } from "react";
import { Link, Redirect } from "react-router-dom";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { register } from "../../actions/authActions";
import { createMessage } from "../../actions/messageActions";
import { InputFieldWithIconPrefix } from "../smallComponents/inputFields";

export const Register = ({ isAuthenticated, register, createMessage }) => {
  const [username, setUsername] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [password2, setPassword2] = useState("");

  const onSubmit = (e) => {
    e.preventDefault();
    if (password != password2) {
      createMessage({ fail: "Passwords do not match" });
    } else {
      const newUser = {
        username,
        email,
        password,
      };
      register(newUser);
    }
  };

  return isAuthenticated ? (
    <Redirect to="/" />
  ) : (
    <div>
      <section className="section section-login">
        <div className="form-container">
          <div className="row">
            <div className="col s12 m8 offset-m2 l6 offset-l3">
              <div className="card-panel login blue white-text center">
                <h2>Register</h2>
                <form onSubmit={onSubmit}>
                  <InputFieldWithIconPrefix
                    id="username"
                    type="text"
                    name="User Name"
                    value={username}
                    setFunction={setUsername}
                    required={true}
                    iconName="account_box"
                  />
                  <InputFieldWithIconPrefix
                    id="email"
                    type="email"
                    name="Email Address"
                    value={email}
                    setFunction={setEmail}
                    required={true}
                    iconName="email"
                  />
                  <InputFieldWithIconPrefix
                    id="password"
                    type="password"
                    name="Password"
                    value={password}
                    setFunction={setPassword}
                    required={true}
                    minLength="6"
                    iconName="lock"
                  />
                  <InputFieldWithIconPrefix
                    id="password2"
                    type="password"
                    name="Password"
                    value={password2}
                    setFunction={setPassword2}
                    required={true}
                    minLength="6"
                    iconName="lock"
                  />
                  <button
                    id="registerSubmitButton"
                    type="submit"
                    className="btn  #546e7a blue-grey darken-1 waves-effect waves-light"
                  >
                    Register
                  </button>
                  <p>
                    Already have an account?{" "}
                    <Link to="/login" style={{ color: "#FFF" }}>
                      Login
                    </Link>
                  </p>
                </form>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  );
};

Register.propTypes = {
  register: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
});

export default connect(mapStateToProps, { register, createMessage })(Register);
