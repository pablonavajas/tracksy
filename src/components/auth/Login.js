import React from 'react';

const Login = props => {
  const onSubmit = e => {};
  const onChange = e => {};

  const email = '';
  const password = '';

  return (
    <div>
      <section className="section section-login">
        <div className="form-container">
          <div className="row">
            <div className="col s12 m8 offset-m2 l6 offset-l3">
              <div className="card-panel login blue white-text center">
                <h2>Tracksy Login</h2>
                <form action="index.html" onSubmit={onSubmit}>
                  <div className="input-field">
                    <i className="material-icons prefix">email</i>
                    <input
                      type="email"
                      id="email"
                      onChange={onChange}
                      required
                    />
                    <label className="white-text">Email</label>
                  </div>
                  <div className="input-field">
                    <i className="material-icons prefix">lock</i>
                    <input
                      type="password"
                      id="password"
                      onChange={onChange}
                      required
                      minLength="6"
                    />
                    <label className="white-text">Password</label>
                  </div>
                  <input
                    type="submit"
                    value="Login"
                    className="btn btn-large btn-extended grey lighten-4 black-text"
                  />
                </form>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  );
};

export default Login;
