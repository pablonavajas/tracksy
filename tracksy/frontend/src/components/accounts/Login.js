import React, { useState } from "react";
import { Link, Redirect } from "react-router-dom";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { login } from "../../actions/authActions";
import { InputFieldWithIconPrefix } from "../smallComponents/inputFields";

export const Login = ({ isAuthenticated, login }) => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const onSubmit = (e) => {
    e.preventDefault();
    login(username, password);
  };

  return isAuthenticated ? (
    <Redirect to="/" />
  ) : (
    <div>
      <section className="section section-login">
        <div className="form-container">
          <div className="row">
            <div className="col s12 m8 offset-m2 l6 offset-l3">
              <div className="card-panel login blue white-text center">
                <h2>Tracksy Login</h2>
                <form onSubmit={onSubmit}>
                  <InputFieldWithIconPrefix
                    id="username"
                    type="text"
                    name="Username"
                    value={username}
                    setFunction={setUsername}
                    required={true}
                    iconName="account_box"
                  />
                  <InputFieldWithIconPrefix
                    id="password"
                    type="password"
                    name="Password"
                    value={password}
                    setFunction={setPassword}
                    required={true}
                    iconName="lock"
                  />
                  <button
                    className="btn #546e7a blue-grey darken-1 waves-effect waves-light"
                    type="submit"
                    name="action"
                  >
                    Login
                  </button>
                  <p>
                    Don't have an account?{" "}
                    <Link to="/register" style={{ color: "#FFF" }}>
                      Register
                    </Link>
                  </p>
                </form>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  );
};

Login.propTypes = {
  login: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
});

export default connect(mapStateToProps, { login })(Login);
