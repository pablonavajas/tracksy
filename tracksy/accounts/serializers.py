from rest_framework import serializers
from django.contrib.auth.models import User
from django.contrib.auth import authenticate

from .models import Connection

User._meta.get_field('email')._unique = True

class ConnectionSerializer(serializers.ModelSerializer):
    class Meta:
        model = Connection
        fields = '__all__'


class UserSerializer(serializers.ModelSerializer):
    isStartup = serializers.BooleanField(source='info.isStartup', default=False, read_only=True)

    class Meta:
        model = User
        fields = ('id', 'username', 'email', 'isStartup')


class RegisterSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = ('id', 'username', 'email', 'password')
        extra_kwargs = {'password': {'write_only': True}}

    def create(self, validated_data):
        user = User.objects.create_user(
            validated_data['username'],
            validated_data['email'],
            validated_data['password']
        )
        return user

class LoginSerializer(serializers.Serializer):
    username = serializers.CharField()
    password = serializers.CharField()

    def validate(self, data):
        user = authenticate(**data)
        if user and user.is_active:
            return user
        raise serializers.ValidationError("Unable to log in. Please verify username and password.")
