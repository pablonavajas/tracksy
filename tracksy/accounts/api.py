from rest_framework.response import Response
from rest_framework.views import APIView
from rest_framework import permissions
from django.contrib.auth import authenticate
from django.contrib.auth.models import User
from django.core.exceptions import FieldError
from rest_framework import serializers
from knox.models import AuthToken
from .serializers import UserSerializer, ConnectionSerializer
from .models import Info


class UserAPI(APIView):
    # Get user info
    def get(self, request):
        if request.user.is_authenticated:
            serializer = UserSerializer(request.user)
            data = serializer.data
            data['isStartup'] = request.user.info.isStartup
            return Response(data)
        raise serializers.ValidationError("Not Logged In")

    # Register
    def post(self, request):
        try:
            isStartup = request.data.pop('isStartup')
        except KeyError:
            raise FieldError("'isStartup' field not provided")
        serializer_user = UserSerializer(data=request.data)
        serializer_user.is_valid(raise_exception=True)
        user = serializer_user.save()
        Info.objects.create(user=user, isStartup=isStartup)
        token = AuthToken.objects.create(user)[1]
        data = serializer_user.data
        data['token'] = token
        data['isStartup'] = user.info.isStartup
        return Response(data)


class LoginAPI(APIView):
    def post(self, request):
        user = authenticate(**request.data)
        if user is not None:
            token = AuthToken.objects.create(user)[1]
            serializer = UserSerializer(user)
            data = serializer.data
            data['token'] = token
            data['isStartup'] = user.info.isStartup
            return Response(data)
        raise serializers.ValidationError("Incorrect Credentials")


class IsStartUpAPI(APIView):
    permission_classes = [permissions.IsAuthenticated]

    def get(self, request):
        return Response(request.user.info.isStartup)


class ConnectionsAPI(APIView):
    permission_classes = [permissions.IsAuthenticated]

    def get(self, request):
        connections = request.user.connections.all()
        serializer = ConnectionSerializer(connections, many=True)
        return Response(serializer.data)

    def post(self, request):
        connections = request.data['connections']
        connectionIds = []
        for connection in connections:
            connection['user'] = request.user.pk
            serializer = ConnectionSerializer(data=connection)
            if serializer.is_valid():
                serializer.save()
                connectionIds.append(serializer.data['id'])
        return Response({"connectionIds": connectionIds})
