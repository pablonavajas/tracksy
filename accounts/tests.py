from django.test import TestCase
from django.urls import reverse
from rest_framework.test import APITestCase, APIClient
from django.contrib.auth.models import User
from accounts.models import Info, Connection
from accounts.serializers import UserSerializer
from accounts.api import RegisterAPI, LoginAPI


class UserGenerator:
    def __init__(self, name="a", password="123456", isStartup=False, add_to_db=False):
        self.username = name
        self.email = name + "@gmail.com"
        self.password = password
        self.isStartup = isStartup
        self.db_representation = None
        if add_to_db:
            self.add_user_to_User_db()
            self.add_info_relation_for_user()

    def as_list(self):
        d = self.as_dict()
        li = []
        for el in d:
            li.append(el)
        return li

    def as_dict(self):
        d = self.__dict__.copy()
        d.pop('db_representation')
        return d

    def data(self):
        """ No password """
        all = self.as_dict()
        all.pop("password")
        return all

    def credentials(self):
        return {"username": self.username, "password": self.password}

    def add_user_to_User_db(self):
        if self.db_representation is None:
            self.db_representation = User.objects.create_user(
                self.username, self.email, self.password)
        return self.db_representation

    def add_info_relation_for_user(self):
        Info.objects.create(user=self.db_representation, isStartup=False)

class ConnectionGenerator:
    def __init__(self, owner, name="seb", description="soft eng"):
        self.owner = owner
        self.name = name
        self.url = "https://www.linkedin.com/in/" + name + "/"
        self.description = description

    def as_list(self):
        d = self.as_dict()
        li = []
        for el in d:
            li.append(el)
        return li

    def as_dict(self):
        return self.__dict__.copy()


class ModelsTest(TestCase):
    def setUp(self):
        self.user = UserGenerator("David")
        self.db_user = self.user.add_user_to_User_db()

    def test_connection_created_with_user_as_fk(self):
        connection = ConnectionGenerator(owner=self.user.username)
        db_connection = Connection.objects.create(
            user=self.db_user,
            **(connection.as_dict())
        )
        user_connection = self.db_user.connections.get(pk=db_connection.pk)
        for key, value in connection.as_dict().items():
            self.assertEqual(getattr(user_connection, key), value)

    def test_info_created_with_user_as_fk(self):
        self.user.add_info_relation_for_user()
        self.assertEqual(self.db_user.info.isStartup, False)


class AccountsApiTest(APITestCase):
    def setUp(self):
        self.user = UserGenerator("David")
        self.user_db = self.user.add_user_to_User_db()
        self.user.add_info_relation_for_user()

    def test_login_user(self):
        credentials = self.user.credentials()
        url = reverse("knox_login")
        response = self.client.post(url, credentials, format='json')
        data = response.data
        self.assertIsNotNone(data.pop("id"))
        self.assertIsNotNone(data.pop("token"))
        self.assertIsNotNone(data.pop("expiry"))
        self.assertEqual(data, self.user.data())

    def test_register_user(self):
        user = UserGenerator("Andrew")
        url = reverse("register")
        response = self.client.post(url, user.as_dict(), format='json')
        data = response.data
        self.assertIsNotNone(data.pop("id"))
        self.assertIsNotNone(data.pop("token"))
        self.assertEqual(data, user.data())

    def test_ConnectionsAPI_post_adds_list_of_connections_for_user(self):
        client = APIClient()
        """ Login User """
        client.force_authenticate(user=self.user_db)

        """ Setting up data for sending """
        owner = self.user.username
        connecitons = [
            ConnectionGenerator(owner, "Seb", "Front End Engineer").as_dict(),
            ConnectionGenerator(owner, "Pablo", "Database Engineer").as_dict()
        ]

        """ Sending a request """
        url = reverse("connections")
        response = client.post(url, {"connections": connecitons}, format='json')
        connection_ids_list = response.data["connectionIds"]
        self.assertEqual(len(connecitons), len(connection_ids_list))



