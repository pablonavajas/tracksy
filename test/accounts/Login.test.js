import React from "react";
import { shallow, mount } from "enzyme";
import { Login } from "../../tracksy/frontend/src/components/accounts/Login";
import ConnectedLogin from "../../tracksy/frontend/src/components/accounts/Login";
import toJson from "enzyme-to-json";
import "../setupTests";
import { HashRouter as Router } from "react-router-dom";

//store needed for testing mapStateToProps
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
const mockStore = configureMockStore([thunk]);

//only VC Funds can register
describe("Register", () => {
  let isAuthenticated, wrapper, login, createMessage, preventDefault;

  beforeEach(() => {
    preventDefault = jest.fn();
    login = jest.fn();
    createMessage = jest.fn();
  });

  const mockEvent = {
    target: {
      name: "description",
      value: "test",
    },
  };

  it("redirects to home page if isAuthenticated is true ", () => {
    isAuthenticated = true;
    wrapper = shallow(
      <Login isAuthenticated={isAuthenticated} login={login} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("shows the login form ", () => {
    isAuthenticated = false;
    wrapper = shallow(
      <Login isAuthenticated={isAuthenticated} login={login} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("calls login function with correct params", () => {
    wrapper = mount(
      <Router>
        <Login isAuthenticated={isAuthenticated} login={login} />
      </Router>
    );

    wrapper.find("#username").at(1).simulate("change", mockEvent);
    wrapper.find("#password").at(1).simulate("change", mockEvent);
    wrapper.find("form").simulate("submit", { preventDefault });
    expect(login).toHaveBeenCalledWith("test", "test");
  });
});

// Testing MapStateToProps
describe("Loging ", () => {
  let store, wrapper;
  const login = jest.fn();

  beforeAll(() => {
    store = mockStore({ auth: { isAuthenticated: false } });
    wrapper = shallow(<ConnectedLogin store={store} login={login} />);
  });

  it("should show the error value imported from store", () => {
    expect(wrapper.props().children.props.isAuthenticated).toBe(false);
  });
});
