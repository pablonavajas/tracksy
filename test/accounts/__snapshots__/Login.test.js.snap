// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register redirects to home page if isAuthenticated is true  1`] = `
<Redirect
  to="/"
/>
`;

exports[`Register shows the login form  1`] = `
<div>
  <section
    className="section section-login"
  >
    <div
      className="form-container"
    >
      <div
        className="row"
      >
        <div
          className="col s12 m8 offset-m2 l6 offset-l3"
        >
          <div
            className="card-panel login blue white-text center"
          >
            <h2>
              Tracksy Login
            </h2>
            <form
              onSubmit={[Function]}
            >
              <InputFieldWithIconPrefix
                iconName="account_box"
                id="username"
                name="Username"
                required={true}
                setFunction={[Function]}
                type="text"
                value=""
              />
              <InputFieldWithIconPrefix
                iconName="lock"
                id="password"
                name="Password"
                required={true}
                setFunction={[Function]}
                type="password"
                value=""
              />
              <button
                className="btn #546e7a blue-grey darken-1 waves-effect waves-light"
                name="action"
                type="submit"
              >
                Login
              </button>
              <p>
                Don't have an account?
                 
                <Link
                  style={
                    Object {
                      "color": "#FFF",
                    }
                  }
                  to="/register"
                >
                  Register
                </Link>
              </p>
            </form>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
